import_obfuscation.go

```diff
@@ -0,0 +1,572 @@
+				privImports.privatePaths = append(privImports.privatePaths, privPaths...)
+				privImports.privateNames = append(privImports.privateNames, privNames...)
+
+				return hashImport(imp, garbledImports)
+			}
+
+			for i := range am.Imports {
+				am.Imports[i].Pkg = initImport(am.Imports[i].Pkg)
+			}
+			for i := range am.Packages {
+				am.Packages[i] = initImport(am.Packages[i])
+			}
+
+			// move imports that contain another import as a substring to the front,
+			// so that the shorter import will not match first and leak part of an
+			// import path
+			sort.Slice(privImports.privatePaths, func(i, j int) bool {
+				iSlashes := strings.Count(privImports.privatePaths[i], "/")
+				jSlashes := strings.Count(privImports.privatePaths[j], "/")
+				// sort by number of slashes first, then alphabetically
+				if iSlashes == jSlashes {
+					return privImports.privatePaths[i] > privImports.privatePaths[j]
+				}
+				return iSlashes > jSlashes
+			})
+			sort.Slice(privImports.privateNames, func(i, j int) bool {
+				return privImports.privateNames[i] > privImports.privateNames[j]
+			})
+			privImports.privatePaths = dedupStrings(privImports.privatePaths)
+			privImports.privateNames = dedupStrings(privImports.privateNames)
+
+			// no private import paths, nothing to garble
+			if len(privImports.privatePaths) == 0 {
+				continue
+			}
+			// log.Printf("\t== Private imports: %v ==\n", privImports)
+
+			// garble all private import paths in all symbol names
+			lists := [][]*goobj2.Sym{am.SymDefs, am.NonPkgSymDefs, am.NonPkgSymRefs}
+			for _, list := range lists {
+				for _, s := range list {
+					// skip debug symbols, and remove the debug symbol's data to save space
+					if s.Kind >= goobj2.SDWARFINFO && s.Kind <= goobj2.SDWARFLINES {
+						s.Data = nil
+						continue
+					}
+
+					// skip local asm symbols. For some reason garbling these breaks things
+					// add the symbol name to a blacklist, so we don't garble related symbols
+					if s.Kind == goobj2.SABIALIAS {
+						if parts := strings.SplitN(s.Name, ".", 2); parts[0] == "main" {
+							skipPrefixes = append(skipPrefixes, s.Name)
+							skipPrefixes = append(skipPrefixes, `"".`+parts[1])
+							continue
+						}
+					}
+
+					// garble read only static data, but not strings. If import paths are in strings,
+					// that means garbling strings might effect the behavior of the compiled binary
+					if s.Kind == goobj2.SRODATA && s.Data != nil && !strings.HasPrefix(s.Name, "go.string.") {
+						var dataTyp dataType
+						if strings.HasPrefix(s.Name, "type..importpath.") {
+							dataTyp = importPath
+						} else if strings.HasPrefix(s.Name, "type..namedata.") {
+							dataTyp = namedata
+						}
+						s.Data = garbleSymData(s.Data, privImports, garbledImports, dataTyp, &buf)
+
+						if s.Size != 0 {
+							s.Size = uint32(len(s.Data))
+						}
+					}
+					s.Name = garbleSymbolName(s.Name, privImports, garbledImports, &sb)
+
+					for i := range s.Reloc {
```

Comment:
What is a reloc?

Reply:
https://en.wikipedia.org/wiki/Relocation_(computing)

Not specific to Go